pkg/resource/obtenant_manager.go

```diff
@@ -0,0 +1,726 @@
+	specLocalityMap := m.GenerateSpecLocalityMap(m.OBTenant.Spec.Pools)
+	statusLocalityMap := m.GenerateStatusLocalityMap(m.OBTenant.Status.Pools)
+	if !reflect.DeepEqual(specLocalityMap, statusLocalityMap) {
+		m.Logger.Info("debug: locality changed", "specLocalityMap", specLocalityMap,
+			"statusLocalityMap", statusLocalityMap)
+		return true
+	}
+	return false
+}
+
+func (m *OBTenantManager) hasModifiedCharset() bool {
+	specCharset := m.OBTenant.Spec.Charset
+	if specCharset == "" {
+		specCharset = tenant.Charset
+	}
+	if specCharset != m.OBTenant.Status.TenantRecordInfo.Charset{
+		m.Logger.Info("debug: charset changed", "specCharset",specCharset,
+			"statusCharset", m.OBTenant.Status.TenantRecordInfo.Charset)
+		return true
+	}
+	return false
+}
+// ---------- buildTenant function ----------
+
+func (m *OBTenantManager) BuildTenantStatus() (*v1alpha1.OBTenantStatus ,error) {
+	tenantName := m.OBTenant.Spec.TenantName
+	tenantCurrentStatus := &v1alpha1.OBTenantStatus{}
+
+	tenantExist, err := m.TenantExist(tenantName)
+	if err != nil {
+		return nil, err
+	}
+	if !tenantExist {
+		return nil, errors.New(fmt.Sprintf("Tenant not exist, Tenant name: %s", tenantName))
+	}
+	gvTenant, err := m.GetTenantByName(tenantName)
+	if err != nil {
+		return nil, errors.Wrap(err, fmt.Sprint("Cannot Get Tenant Failed When Build Tenant Status", tenantName))
+	}
+	m.Logger.Info("debug: gvTenant", "gvTenant", gvTenant)
+
+	poolStatusList, err := m.BuildPoolStatusList(gvTenant)
+
+
+	if err != nil {
+		return nil, err
+	}
+	tenantCurrentStatus.Status = m.OBTenant.Status.Status
+	tenantCurrentStatus.Pools = poolStatusList
+	tenantCurrentStatus.OperationContext = m.OBTenant.Status.OperationContext
+
+	tenantCurrentStatus.TenantRecordInfo = v1alpha1.TenantRecordInfo{}
+	tenantCurrentStatus.TenantRecordInfo.TenantID = int(gvTenant.TenantID)
+
+	// TODO get whitelist from tenant account
+	whitelist, exists := GlobalWhiteListMap[gvTenant.TenantName]
+	if exists {
+		tenantCurrentStatus.TenantRecordInfo.ConnectWhiteList = whitelist
+	} else {
+		// try update whitelist after the manager restart
+		GlobalWhiteListMap[gvTenant.TenantName] = tenant.DefaultOBTcpInvitedNodes
+		tenantCurrentStatus.TenantRecordInfo.ConnectWhiteList = GlobalWhiteListMap[gvTenant.TenantName]
+	}
+
+	tenantCurrentStatus.TenantRecordInfo.UnitNumber = poolStatusList[0].UnitNumber
+	charset, err := m.GetCharset()
+	if err != nil {
+		return nil, err
+	}
+	tenantCurrentStatus.TenantRecordInfo.Charset = charset
+	tenantCurrentStatus.TenantRecordInfo.Locality = gvTenant.Locality
+	tenantCurrentStatus.TenantRecordInfo.PrimaryZone = gvTenant.PrimaryZone
+	poolList := make([]string, 0)
+	zoneList := make([]string, 0)
+	for _, pool := range tenantCurrentStatus.Pools {
+		poolList = append(poolList, m.GeneratePoolName(pool.ZoneList))
+		zoneList = append(zoneList, pool.ZoneList)
+	}
+	tenantCurrentStatus.TenantRecordInfo.PoolList = strings.Join(poolList, ",")
```

Comment:
It's a little bit repetitive to store these info into TenantRecordInfo since there's already pools stored in it